ProjFS.Mac/PrjFSKextTests/HandleOperationTests.mm

```diff
@@ -0,0 +1,569 @@
+            _,
+            MessageType_KtoU_RecursivelyEnumerateDirectory,
+            testDirVnode.get(),
+            _,
+            _,
+            _,
+            _,
+            _,
+            nullptr));
+}
+
+- (void) testEmptyDirectoryEnumerates {
+    testDirVnode->attrValues.va_flags = FileFlags_IsEmpty | FileFlags_IsInVirtualizationRoot;
+    const int actionCount = 5;
+    kauth_action_t actions[actionCount] =
+    {
+        KAUTH_VNODE_LIST_DIRECTORY,
+        KAUTH_VNODE_SEARCH,
+        KAUTH_VNODE_READ_SECURITY,
+        KAUTH_VNODE_READ_ATTRIBUTES,
+        KAUTH_VNODE_READ_EXTATTRIBUTES
+    };
+    
+    for (int i = 0; i < actionCount; i++)
+    {
+        HandleVnodeOperation(
+            nullptr,
+            nullptr,
+            actions[i],
+            reinterpret_cast<uintptr_t>(context),
+            reinterpret_cast<uintptr_t>(testDirVnode.get()),
+            0,
+            0);
+        XCTAssertTrue(
+            MockCalls::DidCallFunction(
+                ProviderMessaging_TrySendRequestAndWaitForResponse,
+                _,
+                MessageType_KtoU_EnumerateDirectory,
+                testDirVnode.get(),
+                _,
+                _,
+                _,
+                _,
+                _,
+                nullptr));
+        MockCalls::Clear();
+    }
+}
+
+- (void) testNonEmptyDirectoryDoesNotEnumerate {
+    testDirVnode->attrValues.va_flags = FileFlags_IsInVirtualizationRoot;
+    const int actionCount = 5;
+    kauth_action_t actions[actionCount] =
+    {
+        KAUTH_VNODE_LIST_DIRECTORY,
+        KAUTH_VNODE_SEARCH,
+        KAUTH_VNODE_READ_SECURITY,
+        KAUTH_VNODE_READ_ATTRIBUTES,
+        KAUTH_VNODE_READ_EXTATTRIBUTES
+    };
+    
+    for (int i = 0; i < actionCount; i++)
+    {
+        HandleVnodeOperation(
+            nullptr,
+            nullptr,
+            actions[i],
+            reinterpret_cast<uintptr_t>(context),
+            reinterpret_cast<uintptr_t>(testDirVnode.get()),
+            0,
+            0);
+        XCTAssertFalse(MockCalls::DidCallFunction(ProviderMessaging_TrySendRequestAndWaitForResponse));
+    }
+}
+    
+-(void) testWriteFile {
+    // If we have FileXattrData attribute we should trigger MessageType_KtoU_NotifyFilePreConvertToFull to remove it
+    testFileVnode->attrValues.va_flags = FileFlags_IsEmpty | FileFlags_IsInVirtualizationRoot;
+    SetFileXattrData(testFileVnode);
+    HandleVnodeOperation(
+        nullptr,
+        nullptr,
+        KAUTH_VNODE_WRITE_DATA,
+        reinterpret_cast<uintptr_t>(context),
+        reinterpret_cast<uintptr_t>(testFileVnode.get()),
+        0,
+        0);
+    XCTAssertTrue(
+       MockCalls::DidCallFunction(
+            ProviderMessaging_TrySendRequestAndWaitForResponse,
+            _,
+            MessageType_KtoU_HydrateFile,
+            testFileVnode.get(),
+            _,
+            _,
+            _,
+            _,
+            _,
+            nullptr));
+    XCTAssertTrue(
```

Comment:
Similar comment as above, can we check `callSequenceNumber`?

Reply:
If the `callSequenceNumber` work is planned for a follow up PR, could you file a work item for it?

Reply:
I added a sequence check.  It does start at 2.  Previous to this call other mocks are called which up the counter.
We may change our minds on how we want this.  
If we did it function based then we would lose the overall ordering of calls which might be unideal.

Reply:
Instead of validating the specific sequence number, what would you think of an approach like this:

```
int GetFunctionCallSequenceNum(FunctionPointerType function, CHECK_ARGS... checkArgs);
```

And we could have a special value (e.g. -1) defined to be NOT_CALLED.

Then the test could have something like:

```
int AaCallSequence = GetFunctionCallSequenceNum(A, a); // Get sequence number of call to A(a)
int AbCallSequence = GetFunctionCallSequenceNum(A, b); // Get sequence number of call to A(b)
assert (AaCallSequence != NOT_CALLED);
assert (AbCallSequence != NOT_CALLED);
assert (AaCallSequence < AbCallSequence);
```

That way we don't break all the tests if we add another function call in the setup (or somewhere else) and we're still validating the ordering of the functions we care about.

Reply:
\+ @pmj for his thoughts on this as well

Reply:
I think/agree that checking specific sequence number values is much too fragile. I also don't think changing the semantics of the existing `DidCallFunction()` itself is helpful; a lot of the time we don't care about order, so that extra argument will just get in the way in those cases.

My original plan for this feature was something along the following lines:

    XCTAssertTrue(
       MockCalls::DidCallFunctionsInOrder(
            ProviderMessaging_TrySendRequestAndWaitForResponse,
            make_tuple(
                _,
                MessageType_KtoU_NotifyDirectoryPreDelete,
                testDirVnode.get(),
                _,
                _,
                _,
                _,
                _,
                nullptr),
            ProviderMessaging_TrySendRequestAndWaitForResponse,
            make_tuple(
                _,
                MessageType_KtoU_RecursivelyEnumerateDirectory,
                testDirVnode.get(),
                _,
                _,
                _,
                _,
                _,
                nullptr)));

This would assert relative order, but not require *all* called functions to be listed.

Reply:
Ohhh I really like that @pmj.  This was along the lines of a similar solution @wilbaker pinged me.